malinka.el

```diff
@@ -843,28 +854,72 @@ Returns the output of the command as a string or nil in case of error"
         (kill-buffer "*rdm*"))))
   (if (rtags-start-process-unless-running)
       t
-      ;else
-      (malinka--error "Could not find rtags daemon in the system")))
+    ;; else
+    (malinka--error "Could not find rtags daemon in the system")))
 
-(defun malinka--rtags-is-indexing? ()
-  "Check if rtags is currently indexing anything"
-  (let ((result (string-to-number (malinka--rtags-invoke-with "--is-indexing"))))
-    (when (= result 1) t)))
+(defun malinka--async-rtags-invoke-with (callback &rest args)
+  "Invoke rc (rtags executable) with ARGS as arguments.
 
-(defun malinka--rtags-project-loaded? (project)
-  "Check if rtags has loaded PROJECT."
-  (let* ((output (malinka--rtags-invoke-with "-w"))
-         (loadedlist (s-match "\\(.*\\) <=" output))
-         (dir (when loadedlist (nth 1 loadedlist))))
-    (when dir
-      (f-equal? dir (malinka--project-root-directory project)))))
+Returns the output of the command in CALLBACK."
+  (when (malinka--rtags-assert-rdm-runs)
+    (let* ((rc (rtags-executable-find "rc"))
+           (cmd (s-join " " (cons rc args))))
+      (when rc
+        (async-shell-command-to-string cmd callback)))))
 
-(defun malinka--rtags-project-known? (project)
+(defun malinka--async-rtags-is-indexing? (callback)
+  "Check if rtags is currently indexing anything asynchronously."
+  (malinka--async-rtags-invoke-with
+   (lambda (result)
+     (funcall callback (= (string-to-number result) 1)))
+   "--is-indexing"))
+
+(defun malinka--async-rtags-project-known? (project callback)
   "Check if rtags knows about PROJECT."
-  (let* ((output       (malinka--rtags-invoke-with "-w"))
-         (project-root (malinka--project-root-directory project))
-         (loadedlist (s-match (format "%s.*" project-root) output)))
-    loadedlist))
+  (malinka--async-rtags-invoke-with
+   (lambda (output)
+     (let* ((project-root
+             (expand-file-name (malinka--project-root-directory project)))
+            (loadedlist (or
+                         (s-match project-root output)
+                         (s-match (format "%s.*" project-root) output))))
+       (funcall callback loadedlist)))
+   "-w"))
+
+(defun malinka--async-rtags-project-loaded? (project callback)
+  "Check if rtags has loaded PROJECT."
+  (malinka--async-rtags-invoke-with
+   (lambda (output)
+     (let* ((loadedlist (s-match "\\(.*\\) <=" output))
+            (dir (when loadedlist (nth 1 loadedlist))))
+       (when dir
+         (funcall
+          callback (f-equal? dir (malinka--project-root-directory project))))))
+   "-w"))
+
+(defun async-shell-command-to-string (command callback)
+  "Execute shell command COMMAND asynchronously in the
+  background.
```

Comment:
why is the `background` in its own line here and why unlike the following lines it has 2 spaces before it?